src/tests/auth/test_auth_gitlab.py

```diff
@@ -0,0 +1,202 @@
+        mock_do_update_groups.reset_mock()
+
+        config.authenticator.user_states["user@test.com"]["updating"] = True
+        self.assertEqual(True, config.authenticator.is_active("user@test.com", mock_request_handler))
+        self.assertEqual(False, mock_do_update_groups.called, "do not call parallel updated")
+        mock_do_update_groups.reset_mock()
+
+        mock_time.return_value = 10000.01 + 80*2 + 1  # stale request
+        self.assertEqual(True, config.authenticator.is_active("user@test.com", mock_request_handler))
+        self.assertEqual(True, mock_do_update_groups.called, "parallel but stale")
+        mock_do_update_groups.reset_mock()
+        config.authenticator.user_states = copy.deepcopy(saved_state)
+        mock_time.return_value = 10000.01
+
+        config.authenticator.user_states["user@test.com"]['updated'] = now  # gitlab info updated
+        config.authenticator.user_states["user@test.com"]['updating'] = False
+        self.assertEqual(True, config.authenticator.is_active("user@test.com", mock_request_handler))
+        self.assertEqual(False, mock_do_update_groups.called, "do not update gitlab because ttl not expired")
+        mock_do_update_groups.reset_mock()
+
+        mock_time.return_value = 10000.01 + 81
+        self.assertEqual(True, config.authenticator.is_active("user@test.com", mock_request_handler))
+        self.assertEqual(True, mock_do_update_groups.called, "ttl expired")
+        mock_do_update_groups.reset_mock()
+        config.authenticator.user_states = copy.deepcopy(saved_state)
+        mock_time.return_value = 10000.01
+
+        # session expire test
+        mock_time.return_value = 10000.01 + 601
+        self.assertEqual(False, config.authenticator.is_active("user@test.com", mock_request_handler), "shoud be expired")
+        self.assertEqual(True, mock_dump_sessions_to_file.called, "dump state to file")
+        mock_dump_sessions_to_file.reset_mock()
+        self.assertIsNone(config.authenticator.user_states.get("user@test.com"), "removed from state")
+        self.assertListEqual([], config.authenticator.get_groups("user@test.com"))
+        config.authenticator.user_states = copy.deepcopy(saved_state)
+        mock_time.return_value = 10000.01
+
+        # test clean expire
+        mock_time.return_value = 10000.01 + 601
+        config.authenticator.clean_expired_sessions()
+        self.assertIsNone(config.authenticator.user_states.get("user@test.com"))
+        config.authenticator.user_states = copy.deepcopy(saved_state)
+        mock_time.return_value = 10000.01
+
+        os.close(tmp[0])
+        os.unlink(tmp[1])
+
+    @patch('time.time', mock_time)
+    @patch('auth.auth_gitlab.GitlabOAuthAuthenticator.do_update_user', mock_do_update_user)
+    @patch('auth.auth_gitlab.GitlabOAuthAuthenticator.do_update_groups', mock_do_update_groups)
+    def test_gitlab_oauth_user_read_scope(self):
+        now = time.time()
+
+        state = {
```

Comment:
I think the previous test method has pretty similar state/config. May be you could extract common part to a separate variable/method?

Reply:
Not exactly the same, there is difference. In case of 3-4 methods may be its ok to move base value to separate method and extend it in test, but in this case its just extra complexity